### DISCUSSION TITLE
Improving 2.0 texture retrieval speeds

### DISCUSSION THREAD

**Discussion by thnewlands** [Voltstro-Studios/UnityWebBrowser#74] [Requests]
I've been looking at 2.0 recently and I've found the performance to be slightly slower than the 1.6.3 release. 

It seems like the biggest issue is the garbage accrued between frames when locking between threads when I run it in the profiler. I saw that there's a lot of instrumentation for profiling it and solutions for the problem. Introducing tearing works in most cases but if it slows down too much it can be distracting. 

I wonder if the image retrieval process would run faster if the whole thing ran on the main thread? That way the tasks would always be in sync. The reason I bring it up is it seems like you're already having to depend on the main thread for the Texture2D.Apply call. I tried messing around with turning the process into a coroutine started in BaseUwbClientManager but I'm still learning UniTask so I didn't make a huge amount of progress. 

The other thing I found is that just changing https://github.com/Voltstro-Studios/UnityWebBrowser/blob/ef2e151baabd905ee59aa6e830c9a69853736420/src/Packages/UnityWebBrowser/Runtime/Core/BaseUwbClientManager.cs#L39 this line from FixedUpdate to Update seemed to smooth my experience out a bit. Is there a tradeoff with stability involved with using Update rather than FixedUpdate or is there a good number I should be using for my timestep?

**Comment by thnewlands** [comment]
Update: I was able to test it out by just adding CommunicationsManager.GetPixels() to LoadTextureData() in WebBrowserClient and it is absolutely slower to run it on the main thread. lol.

**Reply by thnewlands** [reply]
I never unlocked the get method for pixels though...

**Reply by Voltstro** [reply]
Yes, this is why I get the pixels on a seperate thread, since it is blocking.

**Comment by Voltstro** [comment]
I think I mainly just used `FixedUpdate` to lock texture updating at 30 fps.

You can get some more performance by switching to using pipes for the communication layer, as well as enabling performance mode (with a tradeoff of having possible screen tearing and no resizing).

**Reply by thnewlands** [reply]
Hadn't considered using pipes! But resizing was a big part of the appeal for upgrading so I need to think about it some more.

**Reply by thnewlands** [reply]
This may be naive but I think the next thing I'm going to try is having a "last good frame" concept that's somewhere between the tearing approach and the waiting one.

**Reply by thnewlands** [reply]
I think I'm going to try for double buffering it and seeing if that works.

**Reply by thnewlands** [reply]
I'm pretty convinced double buffering is the best way to go but it's a little outside of my wheelhouse so I'm having trouble implementing it right now. So far I've tried adding it to the WebBrowserClient by having it hold two texturedatas -- one called "nexttexturedata" where pixel data goes and then trying to add a switching mechanism to the PixelEventTypeReader. Neither have worked the way I wanted them to but I think that has more to do with the limits of my knowledge around async programming and unfamiliarity with this project. 

Can you think of an easy way to implement double buffering in the current architecture? I think you might be able to get rid of performance mode if you got it working.

**Reply by thnewlands** [reply]
OH wow actually I think I got it. You should be able to see it working here: https://github.com/thnewlands/UnityWebBrowser/commit/832d8c53e6aa8acac71b9863489857173a7630de

I can make a pull request if you want @Voltstro or feel free to just copy it into the main project w/ the style you think is appropriate. It could be improved by actually double buffering instead of just running a frame behind but for the most part gets around the stalling problem from before.

**Reply by thnewlands** [reply]
Now that the tearing / stalling is theoretically fixed is there a way to get it to refresh as quickly as it was refreshing in 1.6?

**Reply by thnewlands** [reply]
Pipes seems like a good way to put on the gas but I can't get it running right now. 

I get 
```
[Web Browser]: [Engine]: Error setting up IPC!
UnityEngine.Logger:LogError (string,object)
UnityWebBrowser.Logging.DefaultUnityWebBrowserLogger:Error (object) (at UnityWebBrowser/src/Packages/UnityWebBrowser/Runtime/Logging/DefaultUnityWebBrowserLogger.cs:32)
UnityWebBrowser.Logging.ProcessLogHandler:HandleOutputProcessLog (object,System.Diagnostics.DataReceivedEventArgs) (at UnityWebBrowser/src/Packages/UnityWebBrowser/Runtime/Logging/ProcessLogHandler.cs:49)
System.Threading._ThreadPoolWaitCallback:PerformWaitCallback ()
```

**Reply by thnewlands** [reply]
I'm using the version of pipes in the latest version of the project btw

**Reply by Voltstro** [reply]
A PR for the double buffer would be easier please.

I'll have a look at pipes, seeing the same issue, something probs broken with the trimmer again.

**Reply by Voltstro** [reply]
Ok pipes issue should be fixed now.

**Reply by thnewlands** [reply]
I'll prep it a little bit and then get that to you! Thanks for the pipes fix!

**Reply by thnewlands** [reply]
Done! I'm still getting the same error with pipes though.
```
[Web Browser]: [Engine]: Error setting up IPC!
UnityEngine.Logger:LogError (string,object)
UnityWebBrowser.Logging.DefaultUnityWebBrowserLogger:Error (object) (at UnityWebBrowser/src/Packages/UnityWebBrowser/Runtime/Logging/DefaultUnityWebBrowserLogger.cs:32)
UnityWebBrowser.Logging.ProcessLogHandler:HandleOutputProcessLog (object,System.Diagnostics.DataReceivedEventArgs) (at UnityWebBrowser/src/Packages/UnityWebBrowser/Runtime/Logging/ProcessLogHandler.cs:49)
System.Threading._ThreadPoolWaitCallback:PerformWaitCallback ()
```

**Reply by thnewlands** [reply]
I wonder if I misconfigured something? Do I need to run your powershell script again to rebuild anything?

**Reply by thnewlands** [reply]
```
    Directory: E:\Projects\Unity\WadingBrowser\PROJECT\UnityWebBrowser\src\Packages\UnityWebBrowser.Engine.Cef.Win-x64


Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a----         5/10/2022   9:33 PM           1097 LICENSE.md
ConvertFrom-Json : A parameter cannot be found that matches parameter name 'AsHashtable'.
At E:\PROJECT\UnityWebBrowser\src\DevScripts\sync-package-cef-engine.ps1:22
char:68
+ ... geJson = Get-Content $CorePackagePath | ConvertFrom-Json -AsHashtable
+                                                              ~~~~~~~~~~~~
    + CategoryInfo          : InvalidArgument: (:) [ConvertFrom-Json], ParameterBindingException
    + FullyQualifiedErrorId : NamedParameterNotFound,Microsoft.PowerShell.Commands.ConvertFromJsonCommand

Cannot index into a null array.
At E:\PROJECT\UnityWebBrowser\src\DevScripts\sync-package-cef-engine.ps1:23
char:1
+ $CorePackageVersion = $CorePackageJson["version"]
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : NullArray

ConvertFrom-Json : A parameter cannot be found that matches parameter name 'AsHashtable'.
At E:\PROJECT\UnityWebBrowser\src\DevScripts\sync-package-cef-engine.ps1:28
char:68
+ ... Json = Get-Content $EngineCefJsonPath | ConvertFrom-Json -AsHashtable
+                                                              ~~~~~~~~~~~~
    + CategoryInfo          : InvalidArgument: (:) [ConvertFrom-Json], ParameterBindingException
    + FullyQualifiedErrorId : NamedParameterNotFound,Microsoft.PowerShell.Commands.ConvertFromJsonCommand

Cannot index into a null array.
At E:\PROJECT\UnityWebBrowser\src\DevScripts\sync-package-cef-engine.ps1:29
char:1
+ $EngineCefJson["version"] = $CefPackagesVersion
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : NullArray

ConvertFrom-Json : A parameter cannot be found that matches parameter name 'AsHashtable'.
At E:\PROJECT\UnityWebBrowser\src\DevScripts\sync-package-cef-engine.ps1:33
char:78
+ ... = Get-Content $EngineCefLinuxJsonPath | ConvertFrom-Json -AsHashtable
+                                                              ~~~~~~~~~~~~
    + CategoryInfo          : InvalidArgument: (:) [ConvertFrom-Json], ParameterBindingException
    + FullyQualifiedErrorId : NamedParameterNotFound,Microsoft.PowerShell.Commands.ConvertFromJsonCommand

Cannot index into a null array.
At E:\PROJECT\UnityWebBrowser\src\DevScripts\sync-package-cef-engine.ps1:34
char:1
+ $EngineCefLinuxJson["version"] = $CefPackagesVersion
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : NullArray

Cannot index into a null array.
At E:\PROJECT\UnityWebBrowser\src\DevScripts\sync-package-cef-engine.ps1:35
char:1
+ $EngineCefLinuxJson["dependencies"]["dev.voltstro.unitywebbrowser.eng ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : NullArray

ConvertFrom-Json : A parameter cannot be found that matches parameter name 'AsHashtable'.
At E:\PROJECT\UnityWebBrowser\src\DevScripts\sync-package-cef-engine.ps1:39
char:74
+ ... n = Get-Content $EngineCefWinJsonPath | ConvertFrom-Json -AsHashtable
+                                                              ~~~~~~~~~~~~
    + CategoryInfo          : InvalidArgument: (:) [ConvertFrom-Json], ParameterBindingException
    + FullyQualifiedErrorId : NamedParameterNotFound,Microsoft.PowerShell.Commands.ConvertFromJsonCommand

Cannot index into a null array.
At E:\PROJECT\UnityWebBrowser\src\DevScripts\sync-package-cef-engine.ps1:40
char:1
+ $EngineCefWinJson["version"] = $CefPackagesVersion
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : NullArray

Cannot index into a null array.
At E:\PROJECT\UnityWebBrowser\src\DevScripts\sync-package-cef-engine.ps1:41
char:1
+ $EngineCefWinJson["dependencies"]["dev.voltstro.unitywebbrowser.engin ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : NullArray


```


this error in the powershell script might be relevant?

**Reply by Voltstro** [reply]
Yes, you need to run the `src/DevScripts/publish-cef-engine-windows.ps1` Powershell script.

Make sure it is the NEW Powershell version, not the one built into Windows.

Also, when you run a script, make sure your working directory is at the same place the script is located.

**Reply by thnewlands** [reply]
Tried running it with the latest powershell version off github https://github.com/PowerShell/PowerShell/releases/tag/v7.2.3 and got new errors:

```
PS E:\PROJECT\UnityWebBrowser\src\DevScripts> .\sync-package-cef-engine.ps1

    Directory: E:\PROJECT\UnityWebBrowser\src\Packages\UnityWebBrowser.Engine.Cef

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a---           5/10/2022  9:33 PM           1097 LICENSE.md

    Directory: E:\PROJECT\UnityWebBrowser\src\Packages\UnityWebBrowser.Engine.Cef.Linux-x64

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a---           5/10/2022  9:33 PM           1097 LICENSE.md

    Directory: E:\PROJECT\UnityWebBrowser\src\Packages\UnityWebBrowser.Engine.Cef.Win-x64

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a---           5/10/2022  9:33 PM           1097 LICENSE.md
InvalidOperation: E:\PROJECT\UnityWebBrowser\src\DevScripts\sync-package-cef-engine.ps1:29
Line |
  29 |  $EngineCefJson["version"] = $CefPackagesVersion
     |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot index into a null array.

InvalidOperation: E:\PROJECT\UnityWebBrowser\src\DevScripts\sync-package-cef-engine.ps1:34
Line |
  34 |  $EngineCefLinuxJson["version"] = $CefPackagesVersion
     |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot index into a null array.

InvalidOperation: E:\PROJECT\UnityWebBrowser\src\DevScripts\sync-package-cef-engine.ps1:35
Line |
  35 |  $EngineCefLinuxJson["dependencies"]["dev.voltstro.unitywebbrowser.eng …
     |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot index into a null array.

InvalidOperation: E:\PROJECT\UnityWebBrowser\src\DevScripts\sync-package-cef-engine.ps1:40
Line |
  40 |  $EngineCefWinJson["version"] = $CefPackagesVersion
     |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot index into a null array.

InvalidOperation: E:\PROJECT\UnityWebBrowser\src\DevScripts\sync-package-cef-engine.ps1:41
Line |
  41 |  $EngineCefWinJson["dependencies"]["dev.voltstro.unitywebbrowser.engin …
     |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot index into a null array.
     
 ```

**Reply by thnewlands** [reply]
`Get-Content $CorePackagePath | ConvertFrom-Json -AsHashtable` logs this
```
Name                           Value
----                           -----
version                        2.0.0-preview.3
displayName                    Unity Web Browser
author                         Voltstro
unity                          2021.2
name                           dev.voltstro.unitywebbrowser
dependencies                   {com.unity.nuget.newtonsoft-json, com.cysharp.unitask, org.nuget.voltrpc}
samples                        {Basic}
description                    Functional web browser in Unity without any issues or hassle.
```
Which looks good to me? I'm surprised its complaining about not being able to index it.

**Reply by thnewlands** [reply]
Oh weird somehow the package.json files just got erased? I think it happened when I ran all.

**Reply by Voltstro** [reply]
You shouldn't need to run the sync scripts, only the publish scripts which actually build the project and copy files to where they need to be.

The sync scripts are solely there for just updating version numbers in the `package.json` files on releases.

**Reply by Voltstro** [reply]
> Oh weird somehow the package.json files just got erased? I think it happened when I ran all.

Oh, yea, that be what the issue would be from, idk how they got deleted tho.

**Reply by thnewlands** [reply]
That fixed it! The refresh rate of the browser didn't improve that much with pipes unfortunately.

**Reply by Voltstro** [reply]
Hmm, not tho sure on what much more you can do.

Obviously you can just use `Update` instead of `FixedUpdate`, like you said at the start of this discussion.

You could also try changing the delay time to be something lower.
https://github.com/Voltstro-Studios/UnityWebBrowser/blob/6ca16dabad1954305c1c4437d2df163abd6db568/src/Packages/UnityWebBrowser/Runtime/Core/WebBrowserClient.cs#L438

**Reply by thnewlands** [reply]
Ah yeah I tried both of those and it made a slight performance improvement! It seems like it might just be a limitation something lower level. The framerate feels a lot smoother with 1.63 when interacting with video specifically.

**Reply by thnewlands** [reply]
Which is strange because with the double buffer optimization (or using the performance mode) the cpu usage wasn't the issue.

**Reply by thnewlands** [reply]
My guess is it's another issue with locking in the project but I don't really know where to start. Maybe it has something to do with the "threadlock" around GetPixels in WebBrowserCommunicationsManager? I'm not super clear on how PixelsEvents are retrieved through engineProxy.GetPixels()

**Reply by Voltstro** [reply]
It goes from CommunicationsManager.GetPixels -> EngineProxy.GetPixels -> VoltRpc -> Process.

The engine proxy is generated by VoltRpc automatically.

The locking in CommunicationsManager is done because VoltRpc isn't async.

**Reply by thnewlands** [reply]
I'm coming back to my previous theory I think that the issue is too many threads involved. Since there's ultimately three lock statements involved in grabbing the pixels now where there was just one in 1.63. I can't think of a good way to get around this that keeps the flexibility of the architecture you've built with 2.0. But I think that's the way to unlock the speed of rendering and maybe messages if you can figure it out!

**Reply by thnewlands** [reply]
One approach might be to double buffer lower in the architecture to make it available to the main thread as soon as possible.

**Reply by thnewlands** [reply]
![image](https://user-images.githubusercontent.com/4378629/168145364-0542fdad-e161-4105-9b46-dc17b48ba795.png)
Here's a screenshot of one of the threads -- you can see it's also doing a lot of waiting around. I found out you can profile them by clicking "live" in the profiler timeline view.

**Reply by thnewlands** [reply]
If you look at the profiler most of the thread pool workers look like that.

**Reply by thnewlands** [reply]
Kind of looks like GetPixels is getting called by multiple threads too. Is there a way to just keep it on one?

**Reply by Voltstro** [reply]
The `GetPixels` method should only be called by the active thread that is created by running the `PixelDataLoop` task.

The waiting around of the threads looks like something to do with UniTask? Maybe, not too sure. I'll have a look at it when I get home.

**Reply by Voltstro** [reply]
So I did some testing, just manaually creating a thread using .NET's `Thread` class instead of using Tasks for the `PixelDataLoop` method does _slightly_ help with speeds (by about 3ms).

One idea I did come up with was to change the way pixels are requested. Instead of going Unity Request Pixels -> Engine Process -> Unity, we will just do Engine Process -> Unity, basically the same way title/url/progress updates are done. Gonna experiment using this method.

**Reply by thnewlands** [reply]
I think that should help a lot! You could even move the double buffer thing to engine process if blocking is still an issue. Thanks for looking into this.

**Comment by Voltstro** [comment]
The performance at which UWB is getting pixels has significantly been improved now. This was done by updating VoltRpc to be able to use Span to directly access the BufferReader's internal buffer, and by doing a MemCopy of the Span's buffer to the NativeArray's buffer.

It runs at a solid 30 FPS now on TCP comms, which I think is fine for what this project's intended purpose is.
